### ISSUE TITLE
Texture UV gets messed up when importing from Blender

### ISSUE THREAD

**Issue by akjulu** [godotengine/godot#41090]
**Godot version:**
Godot Engine v3.2.2.stable.official

**OS/device including version:**
OS: Windows 10
GPU: Nvidia GeForce RTX 2060 Super with driver version 26.21.14.4108
Issue appears using both GLES2 and GLES3

**Issue description:**
Model imported from Blender seems to have corrupted UV mapping on some polygons.

**Steps to reproduce:**
This is a model I have made in Blender to demonstrate my issue. It's the same quad duplicated multiple times to create a bigger square. On the left is the 512x512 texture I'm using and Blender's UV editor. Each quad has exactly the same UV mapping.

![UVProblem1](https://user-images.githubusercontent.com/3025829/89584987-b8068700-d845-11ea-9138-7f9818d8d054.jpg)

When I export the model as .dae using the Better Collada Exporter plugin and import it to Godot, it ends up looking like this.

![UVProblem2](https://user-images.githubusercontent.com/3025829/89585081-e84e2580-d845-11ea-9c5b-5566f24c8100.jpg)

So, some of the quads look like their UV mapping is corrupted in various ways, even though there shouldn't be any difference between them. I have tried altering the import settings of both the model and the texture, but they don't seem to have any effect on the problem. I have also tried importing the exported .dae back to Blender, and it displays correctly there, so the issue doesn't seem to be with the export plugin.

**Minimal reproduction project:**
[UVProblem.zip](https://github.com/godotengine/godot/files/5038552/UVProblem.zip)

**Comment by Calinou** [comment]
@akjulu Please upload a [minimal reproduction project](https://github.com/godotengine/godot/blob/master/CONTRIBUTING.md#provide-a-simple-example-project) to make this easier to troubleshoot.

**Comment by akjulu** [comment]
I added the issue template and a minimal reproduction project to my original post.

**Comment by TheDuriel** [comment]
Try triangulating before export. Looks like its trying to guess the triangulation of your quads, and gets it wrong. Either on export or import.

**Comment by akjulu** [comment]
I have tried triangulating both through the setting in the export plugin and by manually splitting the quads into triangles, but it doesn't seem to help.

**Comment by TheDuriel** [comment]
Given that collada is an outdated and mostly deprecated format. Does this persist when using GLTF on the newest/or next beta version of Blender?

**Comment by akjulu** [comment]
Importing the model in glTF format fixed the issue. Thank you for the suggestion!
My problem is now solved, but this still seems to be a bug in Godot. Should I close the issue or leave it open?

**Comment by Hevedy** [comment]
I'm here with the same issue and I'm using glTF 2.0

**Comment by kubecz3k** [comment]
We have ported the game to 3.2.2 recently and we have an issue which probably is related. Most of the models which are little more complicated have at least one polygon with wrong uv mappings:
![uv_bug](https://user-images.githubusercontent.com/6129594/91530731-e7238c00-e90b-11ea-9706-1ba02976b8e0.png)
![uv_debug](https://user-images.githubusercontent.com/6129594/91530964-45506f00-e90c-11ea-8abe-5b0674dafbc8.png)
[reproduction_asset.zip](https://github.com/godotengine/godot/files/5140533/reproduction_asset.zip)

**Comment by Calinou** [comment]
I can confirm @kubecz3k's issue on Godot 3.2.2.

Minimal reproduction project with the asset included: [test_uv_import.zip](https://github.com/godotengine/godot/files/5141098/test_uv_import.zip)

**Comment by RevoluPowered** [comment]
The problem is simply expressed like this

**changing is_equal_to is a bad idea and not correct as this case below will fail:**

the reason for this regression is because the uv comparison will fail, you are seeing a symptom of the uv2's being merged incorrectly.

```cpp
Vector3 vec = Vector3(0,0,0.1f)+Vector3(0,0,0.2f)+Vector3(0,0,0.3f)
if(vec == Vector3(0,0,0.6f) )
{ 
// this will never run with that commit 218f38c.
}
```

**Comment by RevoluPowered** [comment]
@Xrayez can you possibly make a test for this in master?

**Comment by lawnjelly** [comment]
> // this will never run with that commit 218f38c.

Surely this is correct behaviour? That's what float math does.

You would normally have another explicit function e.g. `Vector3::is_equal_approx` with an epsilon to deal with cases where this was intended? Overriding basic math operators in this way is asking for trouble, as people will expect them to do what they say.

**Comment by RevoluPowered** [comment]
> > // this will never run with that commit [218f38c](https://github.com/godotengine/godot/commit/218f38c7ecdea970a5e82a48e7782077be4fc248).
> 
> Surely this is correct behaviour? That's what float math does.
> 
> You would normally have another explicit function e.g. `Vector3::is_equal_approx` with an epsilon to deal with cases where this was intended? Overriding basic math operators in this way is asking for trouble, as people will expect them to do what they say.

Good point, then the problem is possibly the other way around, the importer is assuming this is the case?

I honestly think the == should hide the float problems away, but you're right the 'float' should expose it's native type data rather than the corrected data, but the real issue we have now is that is this the ONLY place relying on this behaviour?

**Comment by lawnjelly** [comment]
Yes, imo the problem is in the importer if it is not explicitly using an epsilon. This is ubiquitous in merging floating point vertex data.

**Comment by RevoluPowered** [comment]
> Yes, imo the problem is in the importer if it is not explicitly using an epsilon. This is ubiquitous in merging floating point vertex data.

I can make a PR later on and take a look

**Comment by RevoluPowered** [comment]
https://github.com/godotengine/godot/blob/00949f0c5fcc6a4f8382a4a97d5591fd9ec380f8/editor/import/editor_import_collada.cpp#L741

This is the only thing I can see in the code which would explain the vector3 change issue, IDK though. I will use the sample project and debug when I get some time tomorrow

**Comment by Hevedy** [comment]
What about the non collada format error as I posted and sent an example (?)

**Comment by akien-mga** [comment]
> I can make a PR later on and take a look for the importer bug

Bump :)

CC @aaronfranke too as apparently the issue was caused by 218f38c.

**Comment by Reousa** [comment]
@akien-mga  Will we not be getting a fix in a 3.x version?

**Comment by akien-mga** [comment]
@Reousa The current development version is 4.0, so that's where it should be fixed first. Then the fix can be cherry-picked for the 3.2 branch.

**Comment by aaronfranke** [comment]
Did this bug exist before #18992 was merged? That PR changes Vector3's `==` to use approximate equality, and 218f38c7ecdea970a5e82a48e7782077be4fc248 changes Vector3's `==` back to exact equality. If the bug existed before #18992, then it's not *really* a regression from 218f38c7ecdea970a5e82a48e7782077be4fc248 though the bug still needs fixing. If it didn't, then between #18992 and 218f38c7ecdea970a5e82a48e7782077be4fc248, it's likely that someone (possibly me, though I doubt it) changed code that relied on `==` being approximate. Regardless, the fix is the same, but this could help narrow down the problem.

In a nutshell, we need to find out where in the code it's incorrectly using `a == b` and change it to `a.is_equal_approx(b)`. @RevoluPowered You are 100% correct that in many cases you don't want to use `==`.

**Comment by Reousa** [comment]
> @Reousa The current development version is 4.0, so that's where it should be fixed first. Then the fix can be cherry-picked for the 3.2 branch.

> @Reousa This kind of fix is absolutely cherry-pick-able, so once it's fixed in master (4.0) then it can be fixed in 3.2 as well.

I see, thanks for clarifying! Any tips on where to start if I was to help fix this? Perhaps where the import code is located.

**Comment by aaronfranke** [comment]
@Reousa The first thing I would do is try building commits from the time when this broke, then see what changed. You can first test before #18992, and then if the problem doesn't exist there then you would have to bisect between #18992 and #32477 (192 days of commits), and I think the best way to figure out where it broke is to do a bisect but cherry-pick #32477 on each commit in the bisect and check if it works (since we want to find out what change relied on the behavior of #18992).

**Comment by lawnjelly** [comment]
Don't look at this further, I've found the bug. It is due to another approximate equal change in #18992.

The test for being a duplicate vertex in the set uses the Vector3 < operator, which the PR above has changed to not be a < operator, but some approximate equal version. Changing it back to a < operator fixes the bug.

I can do a PR to fix this, but really we should consider reverting #18992 and other PRs doing approximate equal changes to basic operators. I'll discuss this with reduz later. The problem is that having had this change in for a year, there could be other bugs that are dependent on these changes, so we are in the worst of both worlds.

**Comment by lawnjelly** [comment]
> All approx equal operators need to be reverted, approx equal should only be a special function.

@aaronfranke would you be able to do this (rather than me make a PR)? You know the changes made and it seems you are best placed for reverting.

**Comment by aaronfranke** [comment]
The `<` etc operators were the one place where I wasn't sure what the desired behavior was. The idea is that if the X value is tied except for floating point error, it will instead check Y, same with Y to Z.

@lawnjelly Why does a test for duplicate vertices use `<`? Wouldn't it make more sense to check for duplicate vertices using `==` since the point of the operator is to check if they are equal (e.g. duplicates)? `==` is not approximate in both 3.2 and master.

**Comment by lawnjelly** [comment]
> @lawnjelly Why does a test for duplicate vertices use `<`? Wouldn't it make more sense to check for duplicate vertices using `==` since the point of the operator is to check if they are equal (e.g. duplicates)?

You might well ask. I have no idea, it is how set appears to work (never used it myself). :grin:

It probably would be better to check with equal operator, however the point still stands about the reverting.

**Comment by aaronfranke** [comment]
@lawnjelly I opened #41894, but I'm still highly suspicious of this import code in the first place, especially since 218f38c7ecdea970a5e82a48e7782077be4fc248 doesn't touch the `<` operator but somehow was bisected as where things stopped working. Can you please point out which specific files and lines are responsible for importing and checking the duplicate vertices?

**Comment by lawnjelly** [comment]
> @lawnjelly I opened #41894, but I'm still highly suspicious of this import code in the first place, especially since [218f38c](https://github.com/godotengine/godot/commit/218f38c7ecdea970a5e82a48e7782077be4fc248) doesn't touch the `<` operator but somehow was bisected as where things stopped working. Can you please point out which specific files and lines are responsible for importing and checking the duplicate vertices?

Just try your PR, it fixes the import bug.

Probably because both the equals AND the < operator are used in the texture import. They both probably have to be working correctly for it to function right. I think the < operator was used in the set::find function if I remember correctly, but if you step through debugging an import you should see how it works. I'm really not in a position to explain how the importer works, I didn't write it.

**Comment by akien-mga** [comment]
Fixed by #41894.
